File: core/infrastructure/src/main/java/com/gonezo/multiplatform/audioextraction/infrastructure/resolution/DateResolver.java
package com.gonezo.multiplatform.audioextraction.infrastructure.resolution;

import com.gonezo.multiplatform.audioextraction.domain.model.FieldCandidate;
import com.gonezo.multiplatform.audioextraction.domain.model.ResolvedField;
import com.gonezo.multiplatform.audioextraction.domain.schema.FieldSchema;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.format.DateTimeParseException;
import java.util.List;

final class DateResolver implements FieldResolver {
  @Override
  public boolean supports(FieldSchema schema) {
    return "string".equals(schema.type()) && "date-time".equals(schema.format());
  }

  @Override
  public ResolvedField resolve(String fieldName, FieldSchema schema, List<FieldCandidate> candidates) {
    FieldCandidate candidate = ResolverSupport.bestCandidate(candidates);
    if (candidate == null || candidate.rawValue() == null) {
      List<String> issues = schema.required() ? List.of("missing") : List.of();
      return new ResolvedField(null, 0D, List.of(), issues);
    }

    String raw = String.valueOf(candidate.rawValue()).trim();
    if (raw.isEmpty()) {
      return new ResolvedField(null, 0D, candidate.evidence(), List.of("invalid_format"));
    }

    String normalized = parseDateTime(raw);
    if (normalized == null) {
      return new ResolvedField(null, 0D, candidate.evidence(), List.of("invalid_format"));
    }

    return new ResolvedField(normalized, ResolverSupport.clamp(candidate.confidence()), candidate.evidence(), List.of());
  }

  private String parseDateTime(String value) {
    try {
      return Instant.parse(value).toString();
    } catch (DateTimeParseException ignored) {
      // try as local date
    }

    try {
      return LocalDate.parse(value).atStartOfDay().toInstant(ZoneOffset.UTC).toString();
    } catch (DateTimeParseException ignored) {
      return null;
    }
  }
}

File: core/src/main/kotlin/com/gonezo/audioextraction/infrastructure/resolution/DateResolver.kt
package com.gonezo.audioextraction.infrastructure.resolution

import com.gonezo.audioextraction.domain.model.FieldCandidate
import com.gonezo.audioextraction.domain.model.ResolvedField
import com.gonezo.audioextraction.domain.schema.FieldSchema
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset

class DateResolver : FieldResolver {
  override fun supports(schema: FieldSchema): Boolean = schema.type == "string" && schema.format == "date-time"

  override fun resolve(fieldName: String, schema: FieldSchema, candidates: List<FieldCandidate>): ResolvedField {
    val candidate = ResolverSupport.bestCandidate(candidates)
    if (candidate == null || candidate.rawValue == null) {
      val issues = if (schema.required) listOf("missing") else emptyList()
      return ResolvedField(null, 0.0, emptyList(), issues)
    }

    val raw = candidate.rawValue.toString().trim()
    if (raw.isEmpty()) return ResolvedField(null, 0.0, candidate.evidence, listOf("invalid_format"))

    val normalized = parseDateTime(raw)
      ?: return ResolvedField(null, 0.0, candidate.evidence, listOf("invalid_format"))

    return ResolvedField(normalized, ResolverSupport.clamp(candidate.confidence), candidate.evidence, emptyList())
  }

  private fun parseDateTime(value: String): String? {
    return runCatching { Instant.parse(value).toString() }
      .recoverCatching { LocalDate.parse(value).atStartOfDay().toInstant(ZoneOffset.UTC).toString() }
      .getOrNull()
  }
}
